onnxruntime/python/tools/transformers/convert_beam_search.py

```diff
@@ -0,0 +1,368 @@
+            arguments.extend(args.io_block_list)
+        if args.op_block_list:
+            arguments.append('--op_block_list')
+            arguments.extend(args.op_block_list)
+        if args.node_block_list:
+            arguments.append('--node_block_list')
+            arguments.extend(args.node_block_list)
+        if args.force_fp16_initializers:
+            arguments.append('--force_fp16_initializers')
+
+    convert_gpt2_to_onnx(arguments)
+
+    # Run symbolic shape inference to walk around ORT shape inference issue for subgraph.
+    from onnxruntime.tools.symbolic_shape_infer import SymbolicShapeInference
+    out = SymbolicShapeInference.infer_shapes(onnx.load(args.gpt2_onnx), auto_merge=True, guess_output_rank=False)
+    if out:
+       onnx.save(out, args.gpt2_onnx)
+
+def create_ort_session(model_path, use_gpu):
+    from onnxruntime import SessionOptions, InferenceSession, __version__ as ort_version, GraphOptimizationLevel
+    sess_options = SessionOptions()
+    sess_options.graph_optimization_level = GraphOptimizationLevel.ORT_DISABLE_ALL
+    execution_providers = ['CUDAExecutionProvider', 'CPUExecutionProvider'
+                           ] if use_gpu else ['CPUExecutionProvider']
+
+    ort_session = InferenceSession(model_path, sess_options, providers=execution_providers)
+    return ort_session
+
+def convert_model(args):
+    if os.path.exists(args.gpt2_onnx):
+        print(f"skip convert_to_onnx since path existed: {args.gpt2_onnx}")
+    else:
+        gpt2_to_onnx(args)
+
+    #create_ort_session(args.gpt2_onnx, args.use_gpu)
+
+    global config
+    config = GPT2Config.from_pretrained(args.model_name_or_path, cache_dir=args.cache_dir)
+    print(config)
+
+    eos_token_id = config.eos_token_id
+    pad_token_id = config.eos_token_id
+    vocab_size = config.vocab_size
+
+    model = onnx.load(args.gpt2_onnx)
+    model.graph.name = "gpt2 subgraph"
+    inputs = [
+        "input_ids", "max_length", "min_length", "num_beams", "num_return_sequences", "temperature",
+        "length_penalty", "repetition_penalty", "vocab_mask"
+    ]
+
+    outputs = ["sequences"]
+    if args.output_sequences_scores:
+        outputs.append("sequences_scores")
+    if args.output_token_scores:
+        outputs.append("scores")
+
+    node = helper.make_node('BeamSearch', inputs=inputs, outputs=outputs, name='BeamSearch_GPT2')
+    node.domain = "com.microsoft"
+    node.attribute.extend([
+        helper.make_attribute("eos_token_id", eos_token_id),
+        helper.make_attribute("pad_token_id", pad_token_id),
+        helper.make_attribute("no_repeat_ngram_size", args.no_repeat_ngram_size),
+        helper.make_attribute("early_stopping", 1 if args.early_stopping else 0),
+        helper.make_attribute("body", model.graph),
+    ])
+
+    from onnx import TensorProto
+
+    # graph inputs
+    input_ids = helper.make_tensor_value_info('input_ids', TensorProto.INT32, ['batch_size', 'sequence_length'])
```

Comment:
is this assumption correct? I see gpt2 has int64 when converted to onnx.

Reply:
This is input for beam search op, see the operator spec.
gpt2 subgraph indeed use int64 input and output so data cast is needed.